### PR TITLE
API: Change 'filters' option to be additive

### LINKED ISSUE #3205: [BUG]: removeFilters query param does not work with JSON body
### Describe the bug

The `removeFilters` query parameter does not work for API requests that with a JSON body.

### Steps to reproduce

Create the file `godbolt.json` with the contents:

```json
{
    "source": "short scale3(short x, short y, short z) {\nshort t = 10*y + z + y*x;\nreturn t;\n}",
    "options": {}
}
```

Then run `curl 'https://godbolt.org/api/compiler/g111/compile?removeFilters=intel' -H 'Content-Type: application/json' --data-binary @godbolt.json`

### Expected behavior

The returned assembly should be in ATT syntax.

Note that this works as expected when using

```console
curl 'https://godbolt.org/api/compiler/g111/compile?removeFilters=intel' --data-binary 'short scale3(short x, short y, short z) {short t = 10*y + z + y*x;return t;}'
```

### Reproduction link

Not applicable

### Screenshots

Not applicable

### Operating System

N/A

### Browser version

N/A

### PR DESCRIPTION
**Pull request by gpanders**
This extends the 'addFilters' and 'removeFilters' query params to the
JSON API. This allows individual filters to be added or removed without
changing the existing set of default filters. Probably the most useful
application of this is to use

    removeFilters: ["intel"]

Currently, disabling Intel syntax requires having to copy the existing
default filters table and modifying the 'intel' entry.

Closes #3205.

### CODE REVIEW

**Comment by partouf** [comment]
Thanks for the PR!
We've been discussing this internally and kind of want to go a slightly different way.

We would like the supplied filters to supplement and override the defaults - so if you supply `{intel:false}`, it would be combined with the defaults, but override intel. And that would save us from having to maintain the extra add/removefilters options.

If you change https://github.com/compiler-explorer/compiler-explorer/pull/3206/files#diff-a8a20a119f2bb465d5e7774555d88106cd04c1918da02e7fcc3d50e60542fc59R285 to `filters =  { ...(compiler.getDefaultFilters()), ...(requestOptions.filters || {}) };` we think that would be possible.

Could you change your PR accordingly?

**Comment by gpanders** [comment]
Ok done (I actually like that design better anyway 😃).

I'm having some difficulty with the tests though and I wonder if I could get some help? My understanding is that the tests are using the `fake-for-test` compiler, which is defined in `lib/compilers/fake-for-test.js` and returns an empty object for `getDefaultFilters()`. This should mean that in the "parses options and filters" test the only filters it has are the ones passed to it `{a: true, b: true, c: true}`. However, this test is failing and I'm not sure why.

File: docs/API.md [hunk 1]
```diff
@@ -121,7 +121,9 @@ Execution Only request example:
 ```
 
 The filters are a JSON object with `true`/`false` values. If not supplied,
- defaults are used. If supplied, the filters are used as-is.
+ defaults are used. If supplied, the filters are used as-is. The `addFilters`
+ and `removeFilters` option can be used to add or remove individual filters
+ without affecting the default values.
```
Comment (supergrecko) [inline docs/API.md:126]: Should be updated again to reflect the newest change

**Review by supergrecko — APPROVED** [review]

**Comment by apmorton** [comment]
I would like to note that the line of code partouf suggested came from me, and I wrote it into discord on my phone late at night.

Before this is merged someone needs to confirm that it _actually_ does the expected thing.

I put the `|| {}` bit in to hedge bets since I wasn't sure if it was legal to spread an undefined, but judging from the lint error it may be unnecessary.

Someone should confirm that as well.

Thanks all

**Comment by gpanders** [comment]
I tested it locally and it looks like it works without the `|| {}`, so I'll remove that.

**Comment by partouf** [comment]
this is now live

**Pull request merged**